### PR TITLE
M9-3-3: address fp issue 381

### LINKED ISSUE #381: `M9-3-3`: Identify indirect assignment of member data
### Affected rules
 - `M9-3-3`

### Description

Improve the rule to detect assignment into references or pointers into member data.

### Example
In this example we see a member function which modifies a member variable called `values_`, by iterating through and taking a reference to each element.

```cpp
template <typename U = T>
void fill(const T& val) {
    for (auto& elem : values_) {
      elem = val;
    }
  }
```

### PR DESCRIPTION
**Pull request by knewbury01**
## Description

fixes #381 

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - M9-3-3

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/autosar/src/rules/M9-3-3/MemberFunctionConstIfPossible.ql [hunk 1]
```diff
@@ -57,7 +64,9 @@ class ConstMemberFunctionCandidate extends NonConstMemberFunction {
     this.hasDefinition() and
     // For uninstantiated templates we have only partial information that prevents us from determining
     // if the candidate calls non-const functions. Therefore we exclude these.
-    not this.isFromUninstantiatedTemplate(_)
+    not this.isFromUninstantiatedTemplate(_) and
+    //cannot recommend const if it returns a nonconst reference
```
Comment (rvermeulen) [inline cpp/autosar/src/rules/M9-3-3/MemberFunctionConstIfPossible.ql:68]:
```suggestion
    // Cannot recommend const if it returns a non-const reference.
```

File: cpp/autosar/test/rules/M9-3-3/test.cpp [hunk 2]
```diff
@@ -193,3 +193,24 @@ void test_template() {
...
+    for (auto &elem : values) {
+      elem = val;
+    }
+  }
+  constexpr int &front() noexcept { return values[0]; } // COMPLIANT
+};
+
+void fp_reported_in_381() {
+  Z4 z;
+  int i = z.front();
```
Comment (rvermeulen) [inline cpp/autosar/test/rules/M9-3-3/test.cpp:214]: Missing compliance annotation.

**Review by rvermeulen — CHANGES_REQUESTED** [review]
Suggestion for comment and I think we are missing a compliance annotation in the added test case.

File: cpp/autosar/test/rules/M9-3-3/test.cpp (hunk 2 above)
Comment (knewbury01) [reply cpp/autosar/test/rules/M9-3-3/test.cpp:214]: the rule alerts on functions, so their declarations not the usage (the compliance annotation is on [line 209 ](https://github.com/knewbury01/codeql-coding-standards/blob/knewbury01/fix-381/cpp/autosar/test/rules/M9-3-3/test.cpp#L209) but the usage snip is probably in there as I was checking if template instantiation was a factor of this fp/fn case...

good to check tho! thanks!

**Review by rvermeulen — APPROVED** [review]
LGTM.

**Pull request merged**